rfcs/20200624-pluggable-device-for-tensorflow.md

```diff
@@ -0,0 +1,251 @@
+
+* `PluggableDeviceExecutor` implements [StreamExecutor](https://github.com/tensorflow/tensorflow/blob/e5023a1738cce7efcdf9d87863b85c80ab2f8c9e/tensorflow/stream_executor/stream_executor_pimpl.h#L73) and is built on top of StreamExecutor C API (addressed in [RFC](https://github.com/tensorflow/community/pull/257)). 
+
+* `PluggableDevice Implementation` is inside the TensorFlow plugin, which provides those C functions implementation defined in the StreamExecutor C API.
+
+The pluggable device mechanism contains device discovery and creation process which creates a `PluggableDevice` object and `PluggableDeviceExecutor` object for each pluggable device. 
+
+With the RFC, existing TensorFlow GPU programs can run on a plugged device without the user changing the code. The Diagram 2 describes the workflow of TensorFlow with device plugin, it shows how a simple GPU program runs on the pluggable device.
+<div align="center">
+<img src=20200624-pluggable-device-for-tensorflow/gpu_example.png>
+</div>
+
+### Device Discovery
+
+Upon initialization of TensorFlow, it uses platform independent `LoadLibrary()` to load the dynamic library. The plugin library should be installed to default plugin directory "…python_dir.../site-packages/tensorflow-plugins". The modular tensorflow [RFC](https://github.com/tensorflow/community/pull/77) describes the process of loading plugins. 
+
+During the plugin library initialization, it calls the `SE_ReigsterPlatform()` API to register the stream executor platform (`SE_Platform` struct) to TensorFlow proper. The `SE_ReigsterPlatform()` API is a callback API, part of StreamExecutor C API, which passes necessary information to TensorFlow proper to instantiate a stream executor platform ([se::platform](https://github.com/tensorflow/tensorflow/blob/cb32cf0f0160d1f582787119d0480de3ba8b9b53/tensorflow/stream_executor/platform.h#L93) class) and register to a global object [se::MultiPlatformManager](https://github.com/tensorflow/tensorflow/blob/cb32cf0f0160d1f582787119d0480de3ba8b9b53/tensorflow/stream_executor/multi_platform_manager.h#L82). 
+The stream executor platform must be registered with the name "PluggableDevice".  
+See below code which is an example of registering a PluggableDevice platform with StreamExecutor C API:
+```cpp
+void RegisterPluggableDevicePlatform() {
+  static plugin_id_value = 123;
+  SE_PlatformId id;
+  id.id = &plugin_id_value;
+  int visible_device_count = get_plugin_device_count;
+  SE_Platform* custom_platform = SE_NewPlatform(
+     id, visible_device_count,
+     create_device, create_stream_executor,
+     delete_device, delete_stream_executor);
+  TF_Status* status = TF_NewStatus();
+  std::string name = "PluggableDevice";
+  SE_RegisterPlatform(
+     name.c_str(), name.size(),
+     custom_platform,
+     status);
+}
+
+```
+Use static initialization to register the new platform:
```

Comment:
Should we use static initialization or should the dlopen'ed module expose a standardly-named function we can call to register things?

Reply:
I actually don't know the trade off here. We have been using static initialization for kernels registered through C API, so I was thinking of doing same here. Do you think a standardly-named function would be better?

Reply:
Static initializers have a lot of unclear semantics, so if we can avoid them and do a `dlsym` based thing here, that would be safer.

Reply:
I agree with the reviewers; please keep it simple and handle just initialization (i.e. the how). Let the caller decide _when_ to initialize.

Since the types are visible (i.e. not opaque), another improvement we should do is to let the caller decide _where_ to initialize (e.g. heap, stack etc). The caller just passes you a pointer of the type, and plugin fills in the fields.

BTW was this a typo: `RegisterPluggablePlatform`, was it supposed to be `RegisterPluggableDevicePlatform` ?

Reply:
thanks for the reminder, I will fix the name as RegisterPluggableDevicePlatform